### DISCUSSION TITLE
Serialize enums as a string instead of number

### DISCUSSION THREAD

**Discussion by andriyndev** [serverpod/serverpod#879] [Ideas]
(Copy-pasted from https://github.com/serverpod/serverpod/discussions/715 because it's a separate issue)

I think enum's values should never be serialized by numeric values because if the app's developer edits the enum so that the values of the current elements are shifted, it'll cause nasty consequences. For example, user selects one value, but it's written to the database as another one, and the worse is that they'll see it as they specified it while other users with more recent versions of the app will see it as it was written to the database. It might be especially dangerous if the data represents delicate information like user's religion, political view, sexual orientation, gender etc. And even if the app's developers are aware of this nuance, they still might be susceptible to human factor. And if the mistake will be made, it might be impossible to fix the consequences without interacting with each user who triggered writes of the data to the database because there is no way for the app's developers to know whether the value is what the user actually selected.

**Comment by vlidholt** [comment]
You do have a point here. We will have a think about what we can do here. May not be possible to change before version 2.0 though as it would break a lot of things.

**Comment by tgrushka** [comment]
I like this idea. Would be nice if the developer could configure the property to serialize, such as `name` or `index` in the yaml. Maybe default to `name` (`String`)? For compatibility with legacy databases or other reasons, the developer might also want to be able to use an associated value, like:

```yaml
enum: StreamQuality
serializeWith: resolution
properties:
  resolution: String
values:
  - low, resolution='720x480'
  - medium, resolution='1920x1080'
  - high, resolution='3840x2160'
```

```dart
enum StreamQuality with SerializableEntity {
  low('720x480'),
  medium('1920x1080'),
  high('3840x2160');

  final String resolution;

  const StreamQuality(this.resolution);

  static StreamQuality? fromJson(String resolution) =>
      StreamQuality.values.firstWhereOrNull((q) => q.resolution == resolution);

  @override
  String toJson() => resolution;
}
```

**Comment by lukehutch** [comment]
I made the necessary changes in #1080. See further discussion there.
